File: ios/Classes/CourierFlutterChannel.swift
//
//  CourierFlutterChannel.swift
//  courier_flutter
//
//  Created by Michael Miller on 8/8/24.
//

import Flutter

internal enum CourierFlutterChannel: String {
    case shared = "courier_flutter_shared"
    case client = "courier_flutter_client"
    case events = "courier_flutter_events"
}

internal extension CourierFlutterChannel {
    
    var channel: FlutterMethodChannel? {
        get {
            return UIApplication.shared.makeChannel(id: self.rawValue)
        }
    }
    
}

File: android/src/main/kotlin/com/courier/courier_flutter/CourierFlutterChannel.kt
package com.courier.courier_flutter

import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodChannel
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

internal enum class CourierFlutterChannel(val id: String) {
    SHARED("courier_flutter_shared"),
    CLIENT("courier_flutter_client"),
    EVENTS("courier_flutter_events"),
    SYSTEM("courier_flutter_system"),
}

internal fun CourierFlutterChannel.getChannel(messenger: BinaryMessenger) = MethodChannel(messenger, id)

internal fun CourierFlutterChannel.invokeMethod(messenger: BinaryMessenger, method: String, arguments: Any?) {
    CoroutineScope(Dispatchers.Main).launch {
        getChannel(messenger).invokeMethod(method, arguments)
    }
}